proposals/0448-type-variable-scoping.rst

```diff
@@ -0,0 +1,1167 @@
+   see Figure 7, which we modify here in two ways:
+
+   1. Ignore the ``isInternalTypeVar`` premise, which was done
+      away with by accepted proposal `#128`_.
+
+   #. Change the ``cs = ftv(τ's) \ dom(Γ)`` premise to be ``cs = ftv(τ's)``
+      and ``cs # dom(Γ)``. That is, instead of making the new type variables
+      ``cs`` be only those that are not already in scope, require all the
+      type variables to be fresh (shadowing is possible, but left implicit
+      here).
+
+#. A wildcard ``_`` as a type argument says simply to skip that argument;
+   it does not trigger any behavior associated with partial type signatures.
+   In particular, ``-XPartialTypeSignatures`` is not necessary, and no
+   diagnostic is produced.
+
+#. As with term variables, it is an error to bring the same type variable
+   into scope in two (or more) places within the same pattern.
+
+Examples
+^^^^^^^^
+
+::
+
+  {-# LANGUAGE ScopedTypeVariables #-}
+  data Ex = forall a. MkEx a
+  f2 :: forall b. b -> Ex -> Int
+  f2 y (MkEx @b z) = ...
+
+This is rejected under `#126`_,
+as it appears to insist that the existential
+type packed in ``MkEx`` is the same as the type argument passed to ``f2``.
+On the other hand, this is accepted by the current proposal, allowing the
+existential ``b`` to shadow the ``b`` brought into scope by the ``forall``.
+
+This shadowing behavior mimics what happens with term variables in patterns.
+
+Effects
+^^^^^^^
+
+1. The ability to bind existential variables via a construct such as this
+   is necessary to support the `Explicit Variable Principle`_.
+
+#. The previous proposal `#126`_ followed the paper more closely, bringing into
+   scope only those variables that are not already in scope. However, given that
+   this behavior is triggered only by a ``@``, doing this is in violation of the
+   `Visibility Orthogonality Principle`_. This newer version instead labels all variables as binding sites.
+
+#. Having type variables have the same behavior as term variables with
+   respect to shadowing (and repeated binding) upholds the `Visibility Orthogonality Principle`_. In addition,
+   the fact that type variables are unconditionally brought into scope upholds
+   the `Local Lexical Scoping Principle`_.
+
+#. It may be useful to write a variable occurrence to instantiate a universal
+   argument. This proposal prevents this possibility. We expect a future proposal
+   to remedy this problem, with either a modifier or some symbol. For example,
+   perhaps we would say e.g. ``f (Just @(*a) x) = ...`` to denote an occurrence
+   of already-in-scope type variable ``a``.
+
+Type arguments in lambda patterns
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This is a restatement of accepted proposal `#155`_, as amended by not-yet-accepted
+`#238`_. For motivation, please see `#238`_.
+
+Specification
+^^^^^^^^^^^^^
+
+1. With ``-XTypeAbstractions``, introduce a new form of pattern (cf. The `Haskell 2010 Report`_)::
```

Comment:
I believe that this proposal proposes that `-XTypeAbstractions` not be part of `-XScopedTypeVariables`.

Is this on purpose? It seems unfortunate not there to be a single extension which turns on all the things that this proposal deems to be the right way.

Reply:
Good idea.